[WIP] Two nodes openshift (#1675)
openshift/enhancements: enhancements/two-node-fencing/tnf.md

This will allow the use of a credential scoped to `ConfigMap`s in the `openshift-etcd-operator` namespace, to make the change.

The plan is for this to be changed by one of the nodes during pacemaker initialization. Pacemaker initialization should be initiated by CEO when it detects that the cluster controlPlane topology is set to `DualReplica`.

While set to `External`, CEO will still need to render the configuration for the etcd container in a place where it can be consumed by pacemaker. This ensures that the etcd instance managed by pacemaker can be updated
accordingly in the case of a upgrade event or whenever certificates are rotated.

#### Install Config Changes

In order to initialize pacemaker with valid fencing credentials, they will be consumed by the installer via the installation config and created on the cluster as a cluster secret.

A sample install-config.yaml for `platform: none` type clusters would look like this:
```
apiVersion: v1
baseDomain: example.com
compute:
- name: worker
  replicas: 0
controlPlane:
  name: master
  replicas: 2
metadata:
  name: <cluster-name>
platform:
  none:
    fencingCredentials:
      bmc:

>           address: ipmi://<out_of_band_ip>

Reviewer: We are only going to support redfish here I thought?

Reviewer 2: For TP, redfish only. For future releases, we plan on adding IPMI.
I'll update the example.

          username: <user>
          password: <password>
pullSecret: ''
sshKey: ''
```

For platform baremetal, a valid configuration is quite similar.
```
apiVersion: v1
baseDomain: example.com
compute:
- name: worker
  replicas: 0
controlPlane:
  name: master
  replicas: 2
metadata:
  name: <cluster-name>
platform:
  baremetal:
    fencingCredentials:
      bmc:
          address: ipmi://<out_of_band_ip>
          username: <user>
          password: <password>
    apiVIPs:
      - <api_ip>
    ingressVIPs:
      - <wildcard_ip>
pullSecret: ''
sshKey: ''
```

Unfortunately, Baremetal Operator already has a place to specify bmc credentials. However, providing credentials like this will result in conflicts as both the
Baremetal Operator and the pacemaker fencing agent will have control over the machine state. In short, this example shows an invalid configuration that we must check for

> in the installer.

Reviewer 3: BMO is using these credentials to manage the machine state, unsure why we need another way to provide the same information twice in the install-config.
I think it does make sense to add the fencingCredentials for None platform, and in case of baremetal platform just get the bmc creds from the hosts list and create additional secrets (or some other recourse) for  pacemaker fencing agent (in addition to the secrets in the openshift-machine-api namespace)

> in the installer.

Reviewer: We also have to _not_ add the BMC credentials to the BMH CRs in the installer (i.e. they should come out as `unmanaged`, like they do from the assisted SaaS), otherwise ironic will try to fight the fencing agent.

If we were targeting just the `baremetal` platform, as in previous versions of this enhancement, I think we could reuse the existing fields for different purposes in different topologies.

> in the installer.

Reviewer 2: @eranco74 
1. It doesn't solve the conflict. This is an example of an invalid configuration that must result in the installer throwing an error.
2. This is a good point. CC @mshitrit; Have you run into this in your implementation yet?

As for whether or not we should have the fencing credentials in a separate section or not - it comes down to how we want to present a unified API to the end-user. We have two options:
1. Re-use the API(s) exposed in the baremetal specification and update the installer logic to treat them differently when TNF is detected.
2. Introduce a new API for fencing and report an error to the user if they trying to invoke both API(s) as proposed above.

In **both** cases, we need to update BMO to not enlist the control plane nodes into ironic when running in TNF, so there are no advantages either way here.

The first option reuses an existing API, but at the cost of future flexibility. An extension we'd like to explore down the line would be leveraging the MOM operator to install the secret-store-operator, and allow the fencing API to configure a secret entry from a third-party service by configuring the secret-store operator.

My concern is that if we tie ourselves to the Baremetal API now, any future extensions or restrictions that we'd like to make that are TNF specific also have to be enabled in BMO - otherwise we have divergent behavior masked by a singular API.

The other concern I have with reusing the BMO API is that I feel like the specification between platforms should be portable. BMO accepts more than just fencing specifications for its baremetal hosts. A user that does both UPI and Baremetal installs might find it confusing that the APIs for setting up fencing aren't identical between platforms.

> in the installer.

Reviewer 3: I see the point about ensuring a unified API and avoiding conflicts between BMO and the pacemaker fencing agent. However, duplicating the credentials in install-config doesn’t address the conflict resolution. If the goal is to prevent BMO from managing control plane nodes in TNF mode, we should enforce that at the installer level by ensuring these nodes remain unmanaged, as @zaneb  suggested.

Additionally, the hosts field in the baremetal platform is part of the installer API, not BMO. This means the installer is responsible for interpreting and processing these credentials, not BMO itself. 
If we go with a separate fencing API now, we should make sure it doesn't introduce unnecessary duplication for baremetal platform users who would already be defining BMC credentials.
 
 Would it make sense to:

1. Keep fencingCredentials explicitly separate for None platform.
2. For baremetal, derive fencing credentials from existing `Baremetal.Hosts` definitions (creating separate secrets for pacemaker as needed), while ensuring the nodes are unmanaged by BMO. 
3. Ensure the installer throws an error if a user tries to define both APIs at the same time, preventing misconfiguration.

> in the installer.

Reviewer 2: We are all in consensus on points 1 & 3.

Point 2 is the one with the nuance. I agree that the baremetal platform API belongs to the installer, but it is currently organized to expose the fields of the BareMetalHost CRD. Now we can say that this is an implementation detail that is opaque to the user - but this creates dissonance if a user has leveraged this API to set up a non-TNF cluster before.

For an HA cluster deployed with baremetal, I can specify the bmc credentials for each of my hosts and I end up with that configuration being visible in my BareMetalHost CRDs.

For TNF, we specify one fewer control plane node, and this is no longer true.

Weirder still, let's say we (very likely) add support for compute nodes. Now I can specify my 3 nodes, and some of them will have BareMetalHost configuration entries (the compute nodes), and others won't (the control-plane nodes). This is the same API supplying divergent behavior. 

I think it's simpler to have the control plane fencing be specified in a separate place. That way, the installer logic becomes:
- If the control plane fencing is set up and the control plane topology is TNF, then it creates these secrets in CEO to initialize pacemaker.
- If you try to use the fencingCredentials API for non-TNF, you get an error telling you that fencing setup is incompatible with the current topology.
- If the topology is TNF and you're missing the fencing credentials, you throw an error.
- If you try to supply the BMC information in the BareMetalHost API listing, you get an error explaining that BareMetal operator-managed BMC is incompatible with TNF control plane nodes.

Finally, we will need to add logic in Baremetal operator itself that rejects BMC information for control plane node entries in the BMH CRD.

> in the installer.

Author: >This is a good point. CC @mshitrit; Have you run into this in your implementation yet?

ATM I'm still working on the platform none implementation so I haven't run into it yet, however I did put some thought into that.
My initial intention in the baremetal implementation was to make sure that the fencing credentials are mutually exclusive, so they either exist as part of the BMC or in the new TNF format.
Of course this leans on the assumption that we'll create a new API for the TNF.

upgrades from trying to proceed if a node is unavailable.

#### Authentication Operator Changes
The authentication operator is sensitive to the number of kube-api replicas running in the cluster for
[test stability reasons](https://github.com/openshift/cluster-authentication-operator/commit/a08be2324f36ce89908f695a6ff3367ad86c6b78#diff-b6f4bf160fd7e801eadfbfac60b84bd00fcad21a0979c2d730549f10db015645R158-R163).
To get around this, it runs a [readiness check](https://github.com/openshift/cluster-authentication-operator/blob/2d71f164af3f3e9c84eb40669d330df2871956f5/pkg/controllers/readiness/unsupported_override.go#L58)
that needs to be updated to allow for a minimum of two replicas available when using the `DualReplica` infrastructure topology to ensure test stability.

#### Hosted Control Plane Changes
Two-node clusters are no compatible with hosted control planes. A check will be needed to disallow hosted control planes when the infrastructureTopology is set
to `DualReplica`.

#### OLM Filtering Changes
Layering on top of the enhancement proposal for [Two Node with Arbiter (TNA)](../arbiter-clusters.md#olm-filter-addition), it would be ideal to include

> the `DualReplica` infrastructureTopology as an option that operators can levarage to communicate cluster-compatibility.

Reviewer 4: This sounds good to me. However, I'd just like to call out that OLM itself doesn't implement or even know about the infrastructureTopology annotations. The filtering happens only in the OCP console, and there is validation (I think in CVP?) that ensures that all operators explicitly declare support (or not) for each supported infrastructure feature.

Reviewer 5: Is the ask simply to add another filter to consolefor the `operators.openshift.io/infrastructure-features` annotation? That is a straightforward change.

Reviewer 2: What does the infrastructure-features annotation cover?
The specifically affected field is: the `DualReplicaTopologyMode` enum of the ControlPlaneTopologyMode of the openshift infra config. https://github.com/openshift/api/pull/2196/files#diff-07b264a49084976b670fb699badaca1795027d6ea732a99226a5388104f6174fR146

Reviewer 2: I inherited this discussion point from this [thread](https://github.com/openshift/enhancements/pull/1674/files#diff-e0198fba8283a481d7f7032cae79b58cadd8cfa685ab6ea5d8d836114ceb7182R440) in our sibling topology - Two Node OpenShift with Arbiter.

The outcome that we want is a way for operator authors to mark their operators as explicitly compatible or incompatible with one of these ControlPlaneTopologyModes.

Reviewer 5: @jaypoulz Here is the doc around it:

https://docs.openshift.com/container-platform/4.17/operators/operator_sdk/osdk-generating-csvs.html#osdk-csv-annotations-infra_osdk-generating-csvs

They show up like this in the UI:

<img width="214" alt="Screenshot 2025-02-24 at 3 40 24 PM" src="https://github.com/user-attachments/assets/c29634da-4551-49d9-bc7d-d3f0af89e31e" />

It makes sense to me to add it... @joelanford ?

It looks like `operators.openshift.io/infrastructure-features` is actually deprecated in favor of `features.operators.openshift.io` annotations. I'll need to double check console's implementation here. But either way, the frontend work for adding a new filter for this should be straightforward. cc @jhadvig

Reviewer 5: Another example of how they show up in the UI:
<img width="592" alt="Screenshot 2025-02-24 at 3 44 30 PM" src="https://github.com/user-attachments/assets/345ee71e-62b4-4e79-a48a-3d7037fa83b8" />

Reviewer 2: CC @eggfoobar I think this would exactly cover our use case!

The etcd containers are different in this respect since they will be restarted using Podman, but this will be running as root, as it was under CEO.

#### Platform None vs. Baremetal
One of the major design questions of two-node OpenShift is whether to target support for `platform: none` or `platform: baremetal`. The advantage of selecting `platform: baremetal` is that we can leverage the benefits of deploying an ingress-VIP out of the box using keepalived and haproxy. After some discussion with the metal networking team, it is expected that this might work without modifications as long as pacemaker fencing doesn't remove nodes from the node list so that both keepalived instances are always peers. Furthermore, it was noted that this might be solved more simply without keepalived at all by using the ipaddr2 resource agent for pacemaker to run the `ip addr add` and `ip addr remove` commands for the VIP.
The bottom line is that it will take some engineering effort to modify the out-of-the-box in-cluster networking feature for two-node OpenShift.

> Outside of potentially reusing the networking bits of `platform: baremetal`, we discussed potentially reusing its API for collecting BMC credentials for fencing. In this approach, we'd use the `platform: baremetal` BMC entries would be loaded into BareMetalHost CRDs and we'd extend BMO to initialize pacemaker instead of a new operator. After a discussion with the Baremetal Platform team, we were advised against using the Baremetal Operator as an inventory. Its purpose/scope is provisioning nodes.

Reviewer: This is not at all the reason.
The reason is that HA of the baremetal platform components is very much dependent on a working read-write control plane.
Fencing is absolutely part of the purpose/scope of BMO. However, users have already noted that for a 3-node (compact) cluster failover is too slow, and for a 2-node cluster it provably cannot work at all.

Reviewer 2: I think I misinterpreted the intention of [this comment](https://redhat-internal.slack.com/archives/CFP6ST0A3/p1733384274212569?thread_ts=1733356326.224649&cid=CFP6ST0A3).
This section should explain that the existing BMO-based fencing is insufficient to meet our requirements and that integrating with BMO to reuse the existing APIs that BMO exposes would add a dependency on BMO to behave like a baremetal inventory - which is not its purpose.

Reviewer: The way I see it, acting as a baremetal inventory _is_ part of the purpose of BMO.
It just won't work for this use case.

Reviewer 2: Understood! Thanks for taking the time to clarify! :)

This means that the Baremetal Operator is not initially in scope for a two-node cluster because we don't intend to support compute nodes. However, if this requirement were to change for future business opportunities, it may still be useful to provide the user with an install-time option for deploying the Baremetal Operator.

Given the likelihood of customers wanting flexibility over the footprint and capabilities of the platform operators running on the cluster, the safest path forward is to target TNF clusters on both `platform: none` and platform `platform: baremetal` clusters.

> For `platform: none` clusters, this will require customers to provide an ingress load balancer. That said, if in-cluster networking becomes a feature customers request for `platform: none` we can work with the Metal Networking team to prioritize this as a feature for this platform in the future.

Reviewer: Both an Ingress and an API load balancer.

Reviewer: BTW #1666 is the enhancement proposal for this feature, but it seems to be stalled at the moment.

   error to the cluster admin, such that they can see that their cluster may be at risk of failure if no action is taken to resolve the problem? If we introduce a TNF setup
   operator, this could be one of the ongoing functions of this operator. In our current design, we'd likely need to explore what kinds of errors we can bubble up through
   existing cluster health APIs to see if something suitable can be reused.

5. How do we handle updates to the etcd container?

   Things like certificate rotations and image updates will necessitate updates to the pacemaker-controlled etcd container. We will need to introduce some kind of mechanism
   where CEO can describe the changes that need to happen and trigger an image update. We might be able to leverage
   [podman play kube](https://docs.podman.io/en/v4.2/markdown/podman-play-kube.1.html) to map the static pod definition to a container, but we will need to find
   a way to get CEO to render what would usually be the contents of the static pod config to somewhere pacemaker can see updates and respond to them.

## Test Plan

**Note:** *Section not required until targeted at a release.*

### CI

> The initial release of TNF should aim to build a regression baseline.

Reviewer 6: Is the initial release considered TechPreview or is it GA ?

Author: TechPreview

- Documentation for replacing a failed control-plane node
- Documentation for post-installation fencing validation
- Full test coverage (see CI and QE plans above)
- Backhaul SLI telemetry to track TNF installs
- Document SLOs for the topology
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

### Removing a deprecated feature

- Announce deprecation and support policy of the existing feature
- Deprecate the feature

## Upgrade / Downgrade Strategy

This topology has the same expectations for upgrades as the other variants of OpenShift.
For tech preview, upgrades will only be achieved by redeploying the machine and its workload. However,
fully automated upgrades are a requirement for graduating to GA.

One key detail about upgrades is that they will only be allowed to proceed when both nodes are healthy.
The main challenge with upgrading a 2-node cluster is ensuring the cluster stays functional and consistent
through the reboots of the upgrade. This can be achieved by setting the `maxUnavailable` machines in the
control-plane MachineConfigPool to 1.

Downgrades are not supported outside of redeployment.

## Version Skew Strategy

> The biggest concern with version skew would be incompatibilies between a new version of pacemaker and the currently running resource agents.

Reviewer 6: ```suggestion
The biggest concern with version skew would be incompatibilities between a new version of pacemaker and the currently running resource agents.
```

Reviewer 6: Is there any concern with version skew with Corosync ?

> Upgrades will not atomically replace both the RPM and the resource agent configuration, not are there any guarantees that both nodes will be running

Reviewer 6: ```suggestion
Upgrades will not atomically replace both the RPM and the resource agent configuration, nor are there any guarantees that both nodes will be running
```

Author: @dhensel-rh I think atomically is correct here, as a in something that being done in an atomic way, but the `nor` is a good fix.
Can you please modify the suggestion ?

> the same versions. It's difficult to image a case where such an incompatibily wouldn't be caught during upgrade testing. This will be something to keep

Reviewer 6: ```suggestion
the same versions. It's difficult to imagine a case where such an incompatibility wouldn't be caught during upgrade testing. This will be something to keep
```

    concerns as using CEO, just with a different audience.
- Worth noting that the HA team suggests that there is a plan to adjust the pcs to a full client server architecture which will allow the pcs commands to run without root privileges. A partial faster
  solution may be provided by the HA team for a specific set of commands used by the pcs client.

#### Install Config Changes

In order to initialize pacemaker with valid fencing credentials, they will be consumed by the installer via the installation config and created on the cluster as a cluster secret.

A sample install-config.yaml for `platform: none` type clusters would look like this:
```
apiVersion: v1
baseDomain: example.com
compute:
- name: worker
  replicas: 0
controlPlane:
  name: master
  replicas: 2
metadata:
  name: <cluster-name>
platform:
  none:
    fencingCredentials:
      - hostname: <control-0-hostname>
        address: https://<redfish-api-url>
        username: <username>
        password: <password>
      - hostname: <control-1-hostname>
        address: https://<redfish-api-url>
        username: <username>
        password: <password>
pullSecret: ''

> sshKey: ''

Reviewer 7: @rwsu brought this up for discussion this week, and after some consideration I don't think adding `fencingCredentials` to `platform`  makes sense. `platform` is intended to hold cloud-provider specific information. The definition of "cloud provider" may get a bit squishy when we're talking about `baremetal`, but adding to `none` smells: if there is no cloud provider, why associate information with it?

If I understand correctly and there is no relationship between fencingCredentials and the platform, I would suggest moving `fencingCrednetials` out of both `platform.none` and `platform.baremetal`. Adding it to the `MachinePool` and, initially validating to restrict it only to the `controlPlane` machine pool seems like a good option to me. I believe @eranco74 mentioned in another comment we may eventually add support for compute nodes, so I wonder if this idea is compelling in that context?

Furthermore, I would suggest extending the API to make `fencing` a struct, which could potentially be extended in the future if needed, and `credentials` is a child. Putting it together, the example would look like:

```yaml
controlPlane:
  name: master
  replicas: 2
  fencing:
    credentials:
      - hostname: <control-0-hostname>
        address: https://<redfish-api-url>
        username: <username>
        password: <password>
      - hostname: <control-1-hostname>
        address: https://<redfish-api-url>
        username: <username>
        password: <password>
```

If we decide `controlPlane` is not a good fit, my next suggestion would be to move to the top level.
